File: src/main/java/ch/zxseitz/tbsg/security/UserDetailsImpl.java
package ch.zxseitz.tbsg.security;

import ch.zxseitz.tbsg.model.User;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

public class UserDetailsImpl implements UserDetails {
    private final String username;
    private final String password;
    private final List<GrantedAuthority> authorities;

    public UserDetailsImpl(String username, String password, Collection<GrantedAuthority> authorities) {
        this.username = username;
        this.password = password;
        this.authorities = new ArrayList<>(authorities);
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public String getUsername() {
        return username;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }

    public static UserDetailsImpl create(User user) {
        return new UserDetailsImpl(user.getUsername(), user.getPassword(),
                user.getRoles().stream()
                        .map(SimpleGrantedAuthority::new)
                        .collect(Collectors.toList()));
    }
}

File: server/src/main/java/ch/zxseitz/tbsg/server/security/UserDetailsImpl.kt
package ch.zxseitz.tbsg.server.security;

import ch.zxseitz.tbsg.server.model.User;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

open class UserDetailsImpl(
    private val _username: String,
    private val _password: String,
    private val _authorities: List<GrantedAuthority>
) : UserDetails {
    companion object {
        fun create(user: User): UserDetailsImpl {
            return UserDetailsImpl(user.username, user.password, user.roles)
        }
    }

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return _authorities
    }

    override fun getPassword(): String {
        return _password
    }

    override fun getUsername(): String {
        return _username
    }

    override fun isAccountNonExpired(): Boolean {
        return true
    }

    override fun isAccountNonLocked(): Boolean {
        return true
    }

    override fun isCredentialsNonExpired(): Boolean {
        return true
    }

    override fun isEnabled(): Boolean {
        return true
    }
}
